presto-main/src/main/java/com/facebook/presto/server/QueryResource.java

```diff
@@ -100,13 +110,40 @@ public void getAllQueryInfo(
             proxyResponse(servletRequest, asyncResponse, xForwardedProto, uriInfo);
             return;
         }
-        QueryState expectedState = stateFilter == null ? null : QueryState.valueOf(stateFilter.toUpperCase(Locale.ENGLISH));
-        ImmutableList.Builder<BasicQueryInfo> builder = new ImmutableList.Builder<>();
-        for (BasicQueryInfo queryInfo : dispatchManager.getQueries()) {
-            if (stateFilter == null || queryInfo.getState() == expectedState) {
-                builder.add(queryInfo);
-            }
+
+        int limit = firstNonNull(limitFilter, Integer.MAX_VALUE);
+        if (limit <= 0) {
+            throw new WebApplicationException(Response
+                    .status(BAD_REQUEST)
+                    .type(MediaType.TEXT_PLAIN)
+                    .entity(format("Parameter 'limit' for getAllQueryInfo must be positive. Got %d.", limit))
+                    .build());
+        }
+
+        List<BasicQueryInfo> queries = new ArrayList<>(dispatchManager.getQueries());
+
+        // Filter list by the query state (if specified).
+        if (stateFilter != null) {
+            QueryState expectedState = QueryState.valueOf(stateFilter.toUpperCase(Locale.ENGLISH));
+            queries.removeIf(item -> item.getState() == expectedState);
         }
+
+        // If limit is smaller than number of queries, then ensure that the more recent items are at the front.
+        if (limit < queries.size()) {
+            // Sort: RUNNING - first, then QUEUED, then other non-completed, then FAILED and in each group we sort by create time.
+            queries.sort(
+                    Ordering.<BasicQueryInfo>from((o1, o2) -> Boolean.compare(o1.isRunning(), o2.isRunning()))
```

Comment:
Consider making this a constant?

Reply:
Sorry, consider making what a constant?
And how?
Not a java person here, sorry. :)

Reply:
In `QueryResource`, add a `private static final Comparator<BasicQueryInfo> QUERIES_ORDERING = Ordering.from<BasicQueryInfo>from(...` and move the entire contents of lines 135-139 to be assigned to this new variable.